### PR TITLE
feat(testing): Cypress 10 and component testing support

### LINKED ISSUE #6615: Deprecate Storybook Component Testing
## Description
With the release of `cypress@7.0.0` [component testing is no longer considered "experimental"](https://www.cypress.io/blog/2021/04/06/introducing-the-cypress-component-test-runner/). Although it is still in **Alpha release** status, the core Cypress team is rapidly developing tools that support component testing as a first-class citizen within their test runner (SEE: [`@cypress/react`](https://github.com/cypress-io/cypress/tree/develop/npm/react) and [`@cypress/vue`](https://github.com/cypress-io/cypress/tree/develop/npm/vue)).

## Motivation
The [Storybook + Cypress](https://blog.nrwl.io/ui-testing-with-storybook-and-nx-4b86975224c) schematics were a resourceful solution in a time where proper tooling was sorely lacking; however, the landscape is shifting quickly. We ought to create an "end-of-life" process so that Nx users can adopt the native Cypress workflow as soon as possible.

### Key Benefits
- *Simplicity* – No reason to launch a Storybook iFrame when Cypress will offer its own `mount()` function.
- *Performance* – I have not had time to benchmark the two methods, but I speculate that Cypress will be faster.
- *Separation of Concerns* – Storybook for documentation. Cypress for testing.

#### 1. Maximize Testing Support for Component Libraries
- Jest is still an ideal tool for unit testing pure JavaScript, so we do not want to force users to choose between Jest and Cypress.
- Instead, component tests (via Cypress) and unit tests (via Jest) should be able to live side-by-side.
- Therefore, we will need a new target + file extension so that *"test"* target can be reserved for unit testing.
  - `*.spec.{js|ts}` files = Unit tests for Jest
  - `**/*.cy-spec.{jsx,tsx}`= Component tests for Cypress

#### 2. Maximize Framework Support
- Angular = Out of Scope
  - [`@cypress/angular`](https://github.com/cypress-io/cypress/tree/develop/npm/angular) is still work in progress. 
  - Therefore, we would want to leave Storybook + Cypress untouched until it is released on NPM,
- React
  - Begin deprecating all Storybook testing as outlined below

------

#### ~~1. Update Executor in `@nrwl/cypress`~~
~~*Cypress added new CLI commands that are not supported in `@nrwl/cypress:cypress` executor…*~~
> ~~Component tests will no longer be picked up when launching Cypress from cypress open or cypress run. Please use `cypress open-ct` or `cypress run-ct.`~~
> ~~– [Cypress: Component Testing Migration Guide](https://docs.cypress.io/guides/references/migration-guide#4-Use-CLI-commands-to-launch)~~
- ~~**testingType: "component"** should trigger `run-ct` by default~~
- ~~**testingType: "component" + `--watch`** should trigger `open-ct`~~

##### UPDATE 8/29:
The executor works just fine since it is using the NPM API. Also `--watch` trigger is remediated in #6811 

#### 2. Add `@nrwl/react:component-test-configuration`
*Generates default configuration files for Cypress component testing…*

`cypress.json`
```json
{
    "component": {
        "componentFolder": "./src/lib",
        "testFiles": "**/*.cy-spec.{js,jsx,ts,tsx}"
    }
}
```

Other operations might include…
- ADD `cypress` directory with typical sub-directories for fixtures, support, etc.
- ADD `tsconfig.cy-spec.json`, which extends `tsconfig.lib.json` and provides specific overrides.
- UPDATE `workspace.json` with a new target for the specified library (perhaps "ct" or "component")

#### 3. Refactor `@nrwl/react:stories`
*This should **only** create stories for all components declared in a library*

#### 4. Refactor `@nrwl/react:component-cypress-spec`
*This should generate `__componentName__.cy-spec.__fileExt__` as a sibling to Jest test for each component*

#### 5. Update `@nrwl/react:component`
*This should ensure future components have component tests generated by default, alongside the Jest `.spec` files…*
- ADD `__componentName__.cy-spec.__fileExt__`
- UPDATE `component.ts` so that it deletes `*.cy-spec` file if `options.skipTests === true` 

## Out of Scope …but also important
- Find a way to unify test coverage reporting across unit / component tests

## Alternate Implementations
1. **Allow Users to Choose** – Similar to the option to select Karma vs Jest for `unitTestRunner`, we could create Cypress component testing generators as *siblings* of the existing Storybook generators.
2. **Improve Documentation** – As mentioned in #5814 and #5622, the existing workflow is not well documented on Nx.dev.

### LINKED ISSUE #10591: Add support for cypress 10.x
## Current Behavior
- cypress tests do not find cypress.json file with breaking change version

## Expected Behavior
- Expect the e2e test to work and find the cypress.json file
- This is not a regression but an impending bug for upgrading

## Steps to Reproduce
- update cypress to v10.0.2
- run any e2e test 

the failure log below will be visible, where it cannot find the cypress.json even though it is in the specified directory

### Failure Logs

```bash
 wait  - compiling...
       event - compiled client and server successfully in 5.8s (230 modules)
       [ ready ] on http://localhost:9002
       Could not find a Cypress configuration file.
       
       We looked but did not find a cypress.json file in this folder: /home/layton/dev/nx-workspace/apps/static-web-e2e
```

### Environment
```bash
> nx "report"


 >  NX   Report complete - copy this into the issue template

   Node : 14.19.1
   OS   : linux x64
   pnpm : 7.1.9
   
   nx : 14.1.9
   @nrwl/angular : Not Found
   @nrwl/cypress : 14.1.9
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.1.9
   @nrwl/eslint-plugin-nx : 14.1.9
   @nrwl/express : 14.1.9
   @nrwl/jest : 14.1.9
   @nrwl/js : 14.1.9
   @nrwl/linter : 14.1.9
   @nrwl/nest : 14.1.9
   @nrwl/next : 14.1.9
   @nrwl/node : 14.1.9
   @nrwl/nx-cloud : Not Found
   @nrwl/nx-plugin : Not Found
   @nrwl/react : 14.1.9
   @nrwl/react-native : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : 14.1.9
   @nrwl/web : 14.1.9
   @nrwl/workspace : 14.1.9
   typescript : 4.7.3
   ---------------------------------------
   Community plugins:
         @compodoc/compodoc: 1.1.19
         @nx-tools/nx-docker: 2.3.0
         @twittwer/compodoc: 1.8.0
```

### PR DESCRIPTION
**Pull request by barbados-clemens**
## Current Behavior
cypress component testing support is lacking

## Expected Behavior
quality cypress component testing support within nx

### Overview of behavior
```sh
# add component testing to a project
nx g @nrwl/<react|next>:cypress-component-configuration --project=my-react-project

# add component testing to a project and generate component tests
nx g @nrwl/<react|next>:cypress-component-configuration --project=my-react-project --generate-tests

# generate component test for a component
nx g @nrwl/react:component-test --project=my-react-project --component-path=src/lib/my-fancy-component
```

After adding component testing to a project, a new `component-test` target is available to run the tests. `nx component-test my-react-lib`

Cypress component tests and Jest unit tests can live side by side. 

It's not recommended to use cypress component tests for apps, as it's expected apps will not contain any component worth testing. They will primarily be used within libraries. 

Buildable and publishable library support is expected.

Next is not explicitly supported as it's still [alpha by Cypress](https://docs.cypress.io/guides/component-testing/component-framework-configuration#Next-js),

### Migrating to v10
We will not migrate to v10 automatically to prevent breaking e2e tests. but a generator to run manually is provided to aid in the migration when ready. Min cypress supported version is v8

```sh
# convert cypress to v10
nx g @nrwl/cypress:migrate-to-cypress-10
```

This will 
1. update `cypress.json` => `cypress.config.ts` (will handle custom config names as well)
2. update project structure
    - update `support/index.ts` => `support/e2e.ts`
      - and associated imports
    - update `integration/` folder => `e2e/` folder
    - update `.spec.` files to `.cy.` files
3. if plugins are used, import it into the new config file
4. update project configuration
5. install cypress v10

## Progress
| Framework | Project Type | Compiler | Status |
| ----------- | ------------- | --------- | ------ |
| React | library  | babel |  ✅ |
| React | library | swc | ✅ |
| React | app | babel | ✅ |
| React | app | swc | ✅ |

### CODE REVIEW

**Comment by tyler-morrison** [comment]
Dumb question – Do you have a link to the Cypress v10 roadmap?

I couldn't find any notes on the new functionality. Trying to get a feel for how component testing might change from v9 to v10.

**Comment by barbados-clemens** [comment]
@tyler-morrison 
I'm not aware of a docs that explain the differences between v9 and v10 component testing (if any)

I think the usage of it is pretty similar if not the same. but the setup is probably different. at least via looking at the existing docs and seeing how it's set up now, mainly because of the config file change from `cypress.json` to `cypress.config.ts`

File: packages/cypress/src/utils/versions.ts [hunk 1]
```diff
@@ -1,4 +1,13 @@
 export const nxVersion = require('../../package.json').version;
-export const cypressVersion = '^9.1.0';
 export const eslintPluginCypressVersion = '^2.10.3';
 export const typesNodeVersion = '16.11.7';
+/**
+ * TODO(caleb): this is only the mac version, and temp until there is a v10 release
+ */
+export const cypressVersion =
+  'https://cdn.cypress.io/beta/npm/10.0.0/darwin-x64/10.0-release-ae078c74959ddd542ba6ad706613bfec1f07f761/cypress.tgz';
+export const cypressReactVersion = '^5.12.1';
```
Comment (FrozenPandaz) [inline packages/cypress/src/utils/versions.ts:9]: This should move to react?

File: packages/cypress/src/utils/versions.ts [hunk 2]
```diff
@@ -1,4 +1,13 @@
...
+export const cypressWebpackVersion = '^1.8.0';
+export const webpackHttpPluginVersion = '^5.5.0';
+export const swcLoaderVersion = '0.1.15';
```
Comment (FrozenPandaz) [inline packages/cypress/src/utils/versions.ts:12]: This should be there already

File: packages/next/src/generators/component/schema.json [hunk 3]
```diff
@@ -96,6 +96,12 @@
       "type": "boolean",
       "description": "Create component at the source root rather than its own directory.",
       "default": false
+    },
+    "componentTest": {
```
Comment (FrozenPandaz) [inline packages/next/src/generators/component/schema.json:100]: remove

File: packages/react/plugins/component-testing/index.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  return (cypressDevServerConfig, devServerConfig) =>
+    startDevServer({
+      options: cypressDevServerConfig,
+      webpackConfig,
+    });
+}
+
+// TODO(caleb): use the webpack utils to build the config
+//  can't seem to get css modules to play nice when using it 🤔
+function buildBaseWebpackConfig({
```
Comment (FrozenPandaz) [inline packages/react/plugins/component-testing/index.ts:49]: Checkout the storybook

File: packages/storybook/src/generators/cypress-project/cypress-project.ts [hunk 5]
```diff
@@ -70,20 +72,41 @@ export async function cypressProjectGenerator(
...
 
 function addBaseUrlToCypressConfig(tree: Tree, projectName: string) {
-  const cypressConfigPath = `${
-    readProjectConfiguration(tree, projectName).root
-  }/cypress.json`;
-  updateJson(tree, cypressConfigPath, (cypressConfig) => {
-    cypressConfig.baseUrl = 'http://localhost:4400';
-    return cypressConfig;
-  });
+  const projectRoot = readProjectConfiguration(tree, projectName).root;
```
Comment (FrozenPandaz) [inline packages/storybook/src/generators/cypress-project/cypress-project.ts:87]: Talk to @mandarini

File: packages/workspace/src/utils/rules/rename-npm-packages.spec.ts [hunk 6]
```diff
@@ -93,6 +93,7 @@ describe('renameNpmPackages Rule', () => {
 
     tree = await runSchematic('lib', { name: 'library-1' }, tree);
     tree = await runSchematic('lib', { name: 'library-2' }, tree);
+    // dies right here 🤔
```
Comment (FrozenPandaz) [inline packages/workspace/src/utils/rules/rename-npm-packages.spec.ts:96]: This was fixed

File: scripts/depcheck/missing.ts [hunk 7]
```diff
@@ -45,7 +45,13 @@ const IGNORE_MATCHES = {
     'tailwindcss',
   ],
   cli: ['nx'],
-  cypress: ['cypress', '@angular-devkit/schematics', '@nrwl/cypress'],
+  cypress: [
+    'cypress',
+    '@angular-devkit/schematics',
+    '@nrwl/cypress',
+    // migration utilities
+    '@phenomnomnominal/tsquery',
```
Comment (FrozenPandaz) [inline scripts/depcheck/missing.ts:53]: This should be there..

File: scripts/depcheck/missing.ts [hunk 8]
```diff
@@ -45,7 +45,13 @@ const IGNORE_MATCHES = {
     'tailwindcss',
   ],
   cli: ['nx'],
-  cypress: ['cypress', '@angular-devkit/schematics', '@nrwl/cypress'],
+  cypress: [
+    'cypress',
+    '@angular-devkit/schematics',
```
Comment (FrozenPandaz) [inline scripts/depcheck/missing.ts:50]: I don't think this is necessary

File: scripts/depcheck/missing.ts [hunk 9]
```diff
@@ -45,7 +45,13 @@ const IGNORE_MATCHES = {
...
+    '@nrwl/cypress',
```
Comment (FrozenPandaz) [inline scripts/depcheck/missing.ts:51]: This... should not be necessary?

File: scripts/depcheck/missing.ts [hunk 10]
```diff
@@ -82,8 +91,19 @@ const IGNORE_MATCHES = {
     'babel-jest',
     '@angular-devkit/core',
     '@angular-devkit/schematics',
-    '@swc/jest',
-    'babel-jest',
+    // cypress component testing plugin utils, installed dynamically
+    '@cypress/webpack-dev-server',
```
Comment (FrozenPandaz) [inline scripts/depcheck/missing.ts:95]: Checkout what's happening with storybook

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/cypress/src/generators/cypress-component-test-file/schema.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
+{
```
[the remark this answers is not in the document]
Comment (barbados-clemens) [reply packages/cypress/src/generators/cypress-component-test-file/schema.json]: decided to keep.

File: scripts/depcheck/missing.ts (hunk 8 above)
Comment (barbados-clemens) [reply scripts/depcheck/missing.ts:50]: required from older migrations

File: scripts/depcheck/missing.ts (hunk 9 above)
Comment (barbados-clemens) [reply scripts/depcheck/missing.ts:51]: required from plugins which is run-time vs buildtime so it can be safely ignored.

**Comment by tyler-morrison** [comment]
@barbados-clemens I discovered a potential bug with this beta package…

### DESCRIPTION:
The `component-test` tasks hangs when you add the `--watch` flag.

### EXPECTED:
- The tests should run in headed mode.
- Tests should automatically refresh if the test file or source component is modified.

### ACTUAL:
- The Cypress application boots up properly.
- The test runner hangs on "Your tests are loading…"
screenshot of watch flag big

### NOTES:
- The basic `component-test` works great. Apparently the `cypress run-ct` command is unaffected.

**Comment by barbados-clemens** [comment]
@tyler-morrison thanks for giving it a test drive! 😃  haven't run into the watch issue before but I'll investigate some later today to see what it might be.

File: packages/react/src/generators/library/library.spec.ts [hunk 12]
```diff
@@ -6,14 +7,18 @@ import {
...
+import { Schema } from './schema';
+// need to mock cypress otherwise it'll use the nx installed version from package.json
+//  which is v9 while we are testing for the new v10 version
+jest.mock('@nrwl/cypress/src/utils/cypress-version');
 describe('lib', () => {
   let appTree: Tree;
-
+  let mockedInstalledCypressVersion: jest.Mock<
+    ReturnType<typeof installedCypressVersion>
+  > = installedCypressVersion as never;
```
Comment (FrozenPandaz) [inline packages/react/src/generators/library/library.spec.ts:21]: mock return as 10

File: packages/react/src/generators/library/library.ts [hunk 13]
```diff
@@ -100,6 +100,31 @@ export async function libraryGenerator(host: Tree, schema: Schema) {
     tasks.push(jestTask);
   }
 
+  // if (options.addCypress) {
```
Comment (FrozenPandaz) [inline packages/react/src/generators/library/library.ts:103]: Remove

File: scripts/depcheck/missing.ts [hunk 14]
```diff
@@ -82,8 +85,20 @@ const IGNORE_MATCHES = {
...
+    '@nrwl/react',
```
Comment (FrozenPandaz) [inline scripts/depcheck/missing.ts:90]: This shouldn't be here?

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Comment by danr-za** [comment]
truly waiting for that! 👑

File: docs/shared/cypress-plugin.md [hunk 15]
```diff
@@ -11,61 +11,65 @@ Cypress is an e2e test runner built for modern web. It has a lot of great featur
 
 ## Setting Up Cypress
 
-### Generating Applications
+> Info about [Cypress Component Testing can be found here](/cypress/cypress-component-testing)
 
-By default, when creating a new frontend application, Nx will use Cypress to create the e2e tests project.
+If the `@nrwl/cypress` package is not installed, install the version that matches your `@nrwl/workspace` version.
```
Comment (AgentEnder) [inline docs/shared/cypress-plugin.md:16]:
```suggestion
If the `@nrwl/cypress` package is not installed, install the version that matches your `nx` version
```

File: e2e/utils/index.ts [hunk 16]
```diff
@@ -702,7 +702,7 @@ export function checkFilesDoNotExist(...expectedFiles: string[]) {
   expectedFiles.forEach((f) => {
     const ff = f.startsWith('/') ? f : tmpProjPath(f);
     if (exists(ff)) {
-      throw new Error(`File '${ff}' does not exist`);
+      throw new Error(`File '${ff}' does exist`);
```
Comment (AgentEnder) [inline e2e/utils/index.ts:705]:
```suggestion
      throw new Error(`File '${ff}' should not exist`);
```

File: packages/angular/src/generators/ng-add/utilities/file-change-recorder.ts [hunk 17]
```diff
@@ -0,0 +1,44 @@
+import type { Tree } from '@nrwl/devkit';
+import MagicString from 'magic-string';
```
Comment (AgentEnder) [inline packages/angular/src/generators/ng-add/utilities/file-change-recorder.ts:2]: I'm not sure that I like introducing the dependency on this package. It seems like the biggest benefit from it would be generating source maps, and we don't use that. Could this just be a set of simple methods?

I'd argue we don't even really need a "mutable" content, rather appendLeft could easily look like:

```typescript
this.currentContents = this.currentContents.substring(0, index) + insertedString + this.currentContents.substring(index)
```

I'm not really sure that its preferable to change this out, its just something to think about.

File: packages/cypress/src/generators/cypress-component-project/files/cypress/support/component.ts__ext__ [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+//
+// You can change the location of this file or turn off
+// automatically serving support files with the
+// 'supportFile' configuration option.
+//
+// You can read more here:
+// https://on.cypress.io/configuration
+// ***********************************************************
+
+// Import commands.js using ES2015 syntax:
```
Comment (AgentEnder) [inline packages/cypress/src/generators/cypress-component-project/files/cypress/support/component.ts__ext__:16]:
```suggestion
// Import commands.ts using ES2015 syntax:
```

File: packages/cypress/src/generators/cypress-component-project/schema.json [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+  "examples": [
+    {
+      "command": "nx g @nrwl/cypress:cypress-component-project --project=my-cool-lib ",
+      "description": "Add cypress component testing to an existing project named my-cool-lib"
+    }
+  ],
+  "properties": {
+    "project": {
+      "type": "string",
+      "description": "The name of the project to convert",
```
Comment (AgentEnder) [inline packages/cypress/src/generators/cypress-component-project/schema.json:17]: This generator isn't really converting a project, this should read closer to the prompt message

File: packages/cypress/src/generators/cypress-component-project/schema.json [hunk 20]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "http://json-schema.org/schema",
+  "$id": "NxCypressComponentProject",
+  "cli": "nx",
+  "title": "Set up Cypress component test for a project",
```
Comment (AgentEnder) [inline packages/cypress/src/generators/cypress-component-project/schema.json:5]:
```suggestion
  "title": "Set up Cypress component testing for a project",
```

File: packages/cypress/src/generators/cypress-project/files/v9-and-under/src/support/commands.ts__tmpl__ [hunk 21]
```diff
@@ -0,0 +1,33 @@
+// ***********************************************
+// This example commands.js shows you how to
```
Comment (AgentEnder) [inline packages/cypress/src/generators/cypress-project/files/v9-and-under/src/support/commands.ts__tmpl__:2]:
```suggestion
// This example commands.ts shows you how to
```

File: packages/cypress/src/generators/cypress-project/files/v9-and-under/src/support/index.ts__tmpl__ [hunk 22]
```diff
@@ -0,0 +1,17 @@
+// ***********************************************************
+// This example support/index.js is processed and
```
Comment (AgentEnder) [inline packages/cypress/src/generators/cypress-project/files/v9-and-under/src/support/index.ts__tmpl__:2]:
```suggestion
// This example support/index.ts is processed and
```

File: packages/cypress/src/generators/cypress-project/files/v9-and-under/src/support/index.ts__tmpl__ [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+//
+// You can change the location of this file or turn off
+// automatically serving support files with the
+// 'supportFile' configuration option.
+//
+// You can read more here:
+// https://on.cypress.io/configuration
+// ***********************************************************
+
+// Import commands.js using ES2015 syntax:
```
Comment (AgentEnder) [inline packages/cypress/src/generators/cypress-project/files/v9-and-under/src/support/index.ts__tmpl__:16]:
```suggestion
// Import commands.ts using ES2015 syntax:
```

File: packages/cypress/src/generators/migrate-to-cypress-ten/conversion.util.ts [hunk 24]
```diff
@@ -0,0 +1,347 @@
...
+      validFilesEndingsToUpdate.some((e) => path.endsWith(e))
+    ) {
+      updateImports(tree, newPath, oldImportLeafPath, newImportLeafPath);
+    }
+  });
+
+  if (tree.children(oldIntegrationFolder).length === 0) {
+    tree.delete(oldIntegrationFolder);
+  }
+}
```
Comment (AgentEnder) [inline packages/cypress/src/generators/migrate-to-cypress-ten/conversion.util.ts:258]: @FrozenPandaz we should add a `renameFolderInTree` util

File: packages/next/project.json [hunk 25]
```diff
@@ -51,6 +51,11 @@
             "glob": "**/*.d.ts",
             "output": "/"
           },
+          {
+            "input": "packages/next",
+            "glob": "plugins/**/*.html",
+            "output": "/"
+          },
```
Comment (AgentEnder) [inline packages/next/project.json:58]: I'm sure there's a reason, but why is this added? I don't see any changes that would make this needed now.

**Review by AgentEnder — CHANGES_REQUESTED** [review]
@barbados-clemens I didn't have time to review all of this yet, but feel free to move forward when others have reviewed. I left a few minor things, it looks good from what I seen. Most things are teeny-tiny nitpicks.

Only 1 somewhat major thing, but I'm not certain that its the right call. Feel free to ignore it 😅 (MagicString)

File: packages/angular/src/generators/ng-add/utilities/file-change-recorder.ts (hunk 17 above)
Comment (barbados-clemens) [reply packages/angular/src/generators/ng-add/utilities/file-change-recorder.ts:2]: I'll have to defer to @leosvelperez to answer this as this is a part of his work for the angular cli migration to support cypress 10 and not super sure on what his requirements were for this.

File: packages/next/project.json (hunk 25 above)
Comment (barbados-clemens) [reply packages/next/project.json:58]: Ah forgot to remove that when removing the next js support. Thanks for catching it!

File: packages/cypress/src/generators/migrate-to-cypress-ten/conversion.util.ts (hunk 24 above)
Comment (barbados-clemens) [reply packages/cypress/src/generators/migrate-to-cypress-ten/conversion.util.ts:258]: in this case, it makes sense to use the `visitNotIgnoredFiles` since I need to do more than just move the files, but always down for more helper utilities.

File: packages/angular/src/generators/ng-add/utilities/file-change-recorder.ts (hunk 17 above)
Comment (leosvelperez) [reply packages/angular/src/generators/ng-add/utilities/file-change-recorder.ts:2]: @AgentEnder this is not just doing simple string transformations like appendLeft. What `magic-string` allows you to do is to work with a string as if it hasn't changed, so you can keep transforming it based on the original contents and the positions you might have found with the original content. So, that allows us to keep working with the same sourceFile and nodes, using their positions based on the original string, and still be able to continue changing the string without being affected by the positions shifting because the content changes. If we were to do a simple append like the one you suggested, we'd have to keep updating the sourceFile after each modification and re-querying stuff to handle the updated content.

This is pretty handy when dealing with certain AST transformations, like the ones in this PR. The `magic-string` package is used by the Angular DevKit for this same purpose.

**Review by mandarini — APPROVED** [review]
I mainly checked the parts where Storybook is affected. Tested it locally, everything works as expected! Approved by me! :)

File: packages/angular/src/generators/ng-add/utilities/file-change-recorder.ts (hunk 17 above)
Comment (FrozenPandaz) [reply packages/angular/src/generators/ng-add/utilities/file-change-recorder.ts:2]: Could you use `applyChangesToString` for this?

https://github.com/nrwl/nx/blob/d007d37fb4f625fc4854d06d2e083ed778d6a3db/packages/devkit/src/utils/string-change.spec.ts

**Review by FrozenPandaz — APPROVED** [review]

**Comment by rathpc** [comment]
thank you for this! Do you know when this will be in a release?

**Comment by estrudels** [comment]
This is great news that this has been **Merged** which version do we expect this to be released too?

Is it safe to conclude it would be going into version `14.5.0`

Thanks for the effort, we appreciate it! 🙂

**Comment by barbados-clemens** [comment]
@rathpc  @estrudels it should be apart of the 14.5.0 release

**Comment by e-oz** [comment]
Worked like a charm! Thank you so much ❤️

**Comment by rathpc** [comment]
> @rathpc @estrudels it should be apart of the 14.5.0 release

Great news @barbados-clemens! I will try testing out the beta release in the meantime. Do you have a date in mind when the official 14.5.0 will be released?

**Comment by DanielSchiavini** [comment]
I had to rename the files manually, but it worked fine otherwise:
- update integration/ folder => e2e/ folder
- update .spec. files to .cy. files

**Comment by rathpc** [comment]
So I had to end up going with a somewhat custom setup because the default configuration per library which was what I assumed to be the intended setup per several generators did not really work as expected.

For some context, I originally tried having the component test setup love adjacent to the default e2e setup in the automatically created app-e2e folder alongside an app. However if you have any intention of component tests living alongside your components which feels more natural and a practical shift from jest component tests, this setup is not ideal whatsoever and leads to numerous typing issues. The second setup I tried was using a generator to create the config in each library so I could get typing working better and have my component tests adjacent to my components although this added so much config to each and every library it didn't feel worth it. Especially not being able to centralize things like the commands file or types or the component config etc. Additionally if you had any plans of trying to make a target for running these, it would fail if no tests exist in a library as there is no config setting like jest has to pass if no tests are present. Also it wasn't clear how to aggregate those component test results together.

So with all of that said I ended up with the following: cypress folder at the root alongside apps and libs folders with the support folder, fixtures etc in that folder. There is also a tsconfig.json and eslint.config.js to keep things playing nicely. Then at the root are the following files cypress.config.ts, custom tsconfig.cypress.json, vite.config.json, percy.yml, nyc.config.js, nyc.preset.json. My e2e cypress config in apps/app-e2e extends some excluded files from the nyc preset file which I use for coverage, also the vite config uses the same along with the actual nyc config. I have the Percy file because I am also integrating with that plugin for snapshot regression testing. Finally, even though my e2e tests are using webpack as the dev server, a similar setup would just NOT work for component tests and I was forced to go with vite instead. Ironically I think this works way better and I am interested to try to set it up for the e2e tests as well.

Lastly, since this is done at the root I can target a glob to get all component tests from all libs and rely on cypress to parallelize them in ci environments. At least that is what I am hoping for. I haven't gotten far enough to test that yet.

I think it's great we have Cypress 10 support but I am hoping nrwl does a bit more around better config, options and guidelines around how to best set up component tests along with everything else.

Also if anyone from nrwl wants more information on my setup if this sounds like a worthwhile direction feel free to contact me 👍

**Comment by barbados-clemens** [comment]
@rathpc can you explain more on 
> So I had to end up going with a somewhat custom setup because the default configuration per library which was what I assumed to be the intended setup per several generators did not really work as expected.

as cypress component testing generators are built with the idea of adding to an existing project, so any issues you have with the generators not setting that up would be good to know so they can be fixed.

**Comment by rathpc** [comment]
Absolutely (and sorry for the delay!)

So firstly, there are two different generators that almost do the same thing. `@nrwl/cypress:cypress-component-project` and `@nrwl/react:cypress-component-configuration`. This is a bit confusing because they do not result in the same set up configuration. One creates a cypress config with nothing in the defineConfig function, and the other somewhat prepares it.

Secondly, when adding this configuration **per-project** you must ensure that at least one spec file exists with a test before running it because it does not fail gracefully or offer a way to fail gracefully if tests are empty like you can with the Jest executor. If you have a large monorepo with many libraries - odds are you are planning to set up the tooling for them all the same way and be able to leverage nx run or run-many commands. If you did that and any library you set up didn't have any tests yet, the whole command is borked. This was our initial plan because we are hoping to replace all Jest component tests with Cypress component tests and be able to run projects in isolation and collect coverage etc.

So this leads to what I ended up having to do which was initialize a fresh Cypress install at the root of my repo, alongside the `apps` and `libs` folder so that I could be able to scope through all libs, but have the config globally available to initiate the cypress runs. Within each library I had to add a reference to a new tsconfig that I put at the root which was specific for cypress files living in libs so that types would work correctly. The other advantage of keeping cypress at the root is if you intend on having any fixtures, helpers or custom commands that would be shared amongst multiple libs it eliminates the need to duplicate all of that effort in each and every lib you are creating those tests in. For example I created a custom mount command which contains all of our main providers as a wrapper so we can simplify the individual component test mount calls to a single source of truth.

Hopefully this helps explain a bit more 😄

**Comment by barbados-clemens** [comment]
`@nrwl/cypress:cypress-component-project` is the "base" generator that is there to extend for others as desired since it's framework agnostic. i.e. vue plugin wanting to add their own support, they just have to create their own config. 

`@nrwl/react:cypress-component-configuration` adds any react specific needs, such as using the react preset from nx. 

the failing on no tests is a cypress related feature that would have to be added to work the same as jest. I'd recommend not setting up the component testing until you're ready to add tests. or "deactivate" the component tests by renaming the target in the project.json to something like `xcomponent-test` so it won't be picked up via the run-many command. 

as for defining once, and using in multiple places, the cypress and nx teams agree the best way to share a global config would be to define a shared config at the root and then spreading that into each projects config. 
i.e.
```ts
// @ root of repo cypress.shared-config.ts

export default defineConfig({
  // options to globally share across projects
})
```

```ts
// libs/my-fancy-lib/cypress.config.ts

import defaultConfig from '../../cypress.shared-config';

export default defineConfig({
 ...defaultConfig,
 /// project specific options like paths to screenshots/videos
})
```

if you use a single global cypress component testing project for all your components, then you'll missout on any caching of nx and will end up always having to rerun cypress component tests.

if you use a single config but multiple projects, then you'll run into issues that are path reliant such as screenshots/videos/glob patterns that will be per project.

You can optionally make a nx library and create your own cypress commands (and export a config) that can be imported into each project you wish to use it in. 

Here is a showing how to make a command to share, but the same idea would apply to cypress component testing. 

https://www.youtube.com/watch?v=YNH4wBUZvkA

if none of those options can work for you, and you must have a single global project for all your cypress component testing, odds are migrations aren't going to work for you without some extra effort as supporting all permutations of using cypress across a project is quite a task. I can only really support most common setups that I described above.

**Comment by rathpc** [comment]
@barbados-clemens hmm interesting ok well it could be worth trying some more variations when I have time. Ideally I would love to take advantage of the caching but just kept hitting roadblocks so I had to settle to get other work unblocked. Thank you for your response and help!

**Comment by barbados-clemens** [comment]
@rathpc that's totally fair. component testing is very new and our initial vision was using it similar to jest as it's added to a project and run/cacheable like normal unit testing that people were familiar with. 

there are a couple issues with it right now that have been reported around errors getting thrown with different library combinations that I didn't account for. so I understand needed to get it to work today, once I get things more stable for the component testing I can take a look at other ways people have used component testing and see what there might be able to do in regards to smoothing out the transition between different usage patterns.

**Comment by rathpc** [comment]
@barbados-clemens one other thing that is worth mentioning is that I wasn't exactly sure how the following would work (although I didn't attempt it yet):

Assuming I add the component test tooling to each lib and am able to leverage nx run or run-many commands and I can manually aggregate coverage reports from each library into a full coverage report across all libs, is there any way to enable and allow for parallelism with cypress dashboard as well? Keeping everything global seems like it would be the easiest path to allow for this.

I guess the same would apply as a question for jest to be able to run across a matrix of machines in a GitHub workflow, how would the logic work there to be able to determine what runs where if nx is attempting to handle dependency task management?

Even if you choose to just always use run-many and enable parallel runs that way, can you further optimize by running say 3 machines in parallel all running parallel task runs? I am assuming all of these scenarios would require a lot of hand crafted logic to manage and there is nothing out of the box that handles this?

**Comment by wrslatz** [comment]
I've used this [10x Faster CI with Nx and GitHub Actions](https://medium.com/emoteev-blog/10x-faster-ci-with-nx-and-github-actions-9a51fc4e82a6) article as a reference in the past.

If Nx commands supported a `shard` option to automatically split up tasks into groups to spread across CI it would make this a lot easier to manage.

**Comment by rathpc** [comment]
@wrslatz that is an awesome article I will certainly see if I can make use of that! Thank you!

**Comment by barbados-clemens** [comment]
@wrslatz @rathpc I would look into [DTE from Nx Cloud](https://nx.dev/nx-cloud/intro/what-is-nx-cloud), which does this intelligently vs you attempting to load balance all your projects tasks in CI.
